src/lib.rs

```diff
@@ -13,12 +15,29 @@ mod parser;
 pub use error::Error;
 
 pub fn execute(input: &str) -> Result<(String, String), Error> {
-    let (remaining, ast) = parser::program(CompleteStr(input))?;
-    if !remaining.is_empty() {
-        return Err(Error::Parser(format!(
-            "failed to parse program fully, remaining statements:\n{}",
-            remaining
-        )));
+    let res = parser::program(parser::Span::new(CompleteStr(input)));
+    match res {
+        Ok((remaining, ast)) => if !remaining.fragment.is_empty() {
+            Err(Error::Parser(format!(
+                "failed to parse program fully, remaining statements:\n{}",
+                remaining.fragment
+            )))
+        } else {
+            interpreter::evaluate(&ast?)
+        },
+        Err(nom::Err::Failure(nom::simple_errors::Context::Code(
+            position,
+            nom::ErrorKind::Custom(err_code),
+        ))) => {
+            let line = input.lines().nth(position.line as usize - 1).unwrap();
+            let stderr = parser::ErrorCode::from(err_code)
+                .to_string(line, &position.fragment.lines().next().unwrap())
```

Comment:
What type is `position`? If I understand correctly, `position.line` gives the line number. Why are we not passing that?

Reply:
Position is has a type `Span<'a>` which contains a contextual information about the position where the error was caught. The information is `offset`, `line` and remaining str `fragment`, which could not be parsed. All this information is needed sometimes. E.g. `fragment` is used to calculate a cursor under a statement containing error.

Reply:
Ah, line is just the line number, right?